### PR TITLE
Handling SearchServlet Empty Response 

### LINKED ISSUE #78: Search Servlet Check for Empty Response
If too long of a search query is entered, nothing will be returned from the Google Books API. Currently this is not handled as the Servlet is looking for an "items" object even though it doesn't exist, causing it to fail.

Example of query that would generate an error: https://www.googleapis.com/books/v1/volumes?q=reallylongbookthatdoesntexist

### PR DESCRIPTION
**Pull request by antoniolinhart**
# Description
This PR ensures that empty responses are handled on the backend and do not throw null pointer exceptions. The servlet now returns a 204 No Content in the response if there are no books found given a search query.

I have updated all relevant fetches to the /api/search?searchTerm to ensure no exceptions get thrown due to the response and instead set the array of books to be empty, which is handled specially by each implementation.

In addition, I updated the criteria for componentDidUpdate in the ClubSearchList, which was causing the club page to rerender based on the books searched in the previous search query.

# Screenshot of Feature
image

# Linked Issue
#78

### CODE REVIEW

File: capstone-backend/pom.xml [hunk 1]
```diff
@@ -57,11 +57,6 @@
       <artifactId>slf4j-simple</artifactId>
       <version>1.7.25</version>
     </dependency>
-    <dependency>
-      <groupId>com.google.guava</groupId>
-      <artifactId>guava</artifactId>
-      <version>29.0-jre</version>
-    </dependency>
```
Comment (antoniolinhart) [inline capstone-backend/pom.xml]: Confirmed with Steven to figure out that the guava dependency was not causing the build to fail and Guava takes the project quite a long time to deploy locally or remote.

File: capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 2]
```diff
@@ -130,11 +130,13 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
         return;
       }
     } else if (request.getParameter("searchTerm") != null) {
-      volumes.addAll(convertResponseToVolumeData(fullOutput));
-      if (volumes.size() == 0) {
-        response.sendError(HttpServletResponse.SC_NOT_FOUND);
+      // Check for empty response
+      Collection<VolumeData> volumesRetrieved = convertResponseToVolumeData(fullOutput);
+      if(volumesRetrieved == null || volumesRetrieved.size() == 0) {
```
Comment (steven-solar) [inline capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java]: `nit`: space between if and open parenthesis

File: capstone-frontend/src/components/Browse.jsx [hunk 3]
```diff
@@ -14,10 +14,16 @@ export class Browse extends Component {
 
   getSearchedBooks = (searchQuery) => {
     fetch(`/api/search?searchTerm=${searchQuery}`)
-      .then(response => response.json())
+      .then(response => {
```
Comment (steven-solar) [inline capstone-frontend/src/components/Browse.jsx]: could combine this into a one-line ternary if you'd like.

ie `.then(response => response.status === 200 ? response.json() : [])`

File: capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java (hunk 2 above)
Comment (steven-solar) [inline capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java]: nice error checking!

File: capstone-frontend/src/components/SearchBookModal.jsx [hunk 4]
```diff
@@ -33,15 +33,18 @@ export class SearchBookModal extends Component {
...
-        .then(response => response.json())
-        .catch(err => alert(err));
-
-      if (typeof searchResults === 'undefined') {
-        searchResults = [];
-      }
-
-      this.setState({ searchResults, displayBooks: true, fetchingBooks: false })
+      fetch(`/api/search?searchTerm=${searchTerm}&maxResults=${4}`)
+        .then(response => {
```
Comment (steven-solar) [inline capstone-frontend/src/components/SearchBookModal.jsx]: could also make this one line

File: capstone-frontend/src/components/CreateList.jsx [hunk 5]
```diff
@@ -49,21 +49,24 @@ class CreateList extends Component {
...
-        .catch(err => alert(err));
-
-      if (typeof searchResults === "undefined") {
-        searchResults = [];
-      }
-
-      this.setState({ searchResults, displayBooks: true, fetchingBooks: false })
+      fetch(`/api/search?searchTerm=${searchTerm}&maxResults=${4}`)
+        .then(response => {
+          if (response.status === 200) {
```
Comment (steven-solar) [inline capstone-frontend/src/components/CreateList.jsx]: one line

File: capstone-frontend/src/components/ClubSearchList.jsx [hunk 6]
```diff
@@ -66,8 +66,7 @@ export class ClubSearchList extends Component {
   }
 
   componentDidUpdate(prevProps) {
-    if (this.props.searchQuery !== prevProps.searchQuery
-        || this.props.books !== prevProps.books) {
+    if (this.props.books !== prevProps.books) {
```
Comment (steven-solar) [inline capstone-frontend/src/components/ClubSearchList.jsx:69]: Why this change?

File: capstone-frontend/src/components/Browse.jsx [hunk 7]
```diff
@@ -14,10 +14,16 @@ export class Browse extends Component {
...
+        if (response.status === 200) {
+          return response.json();
+        } else {
+          return [];
+        }
+      })
       .then(searchedBooks => this._isMounted && this.setState({ searchedBooks, loading: false }))
       .catch(function (err) {
-        alert(err)
+        console.log(err);
```
Comment (steven-solar) [inline capstone-frontend/src/components/Browse.jsx:21]: Nice on changing alerts to logs

File: capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 8]
```diff
@@ -160,6 +162,9 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     // Get the items element from the output and convert it into the array of book
     // jsons
     JsonElement items = dataInfo.get("items");
+    if (items == null) {
```
Comment (steven-solar) [inline capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java:165]: Nice check

**Review by steven-solar — APPROVED** [review]
Nice fix!

**Review by MauroGuerrero20 — APPROVED** [review]
LGTM

File: capstone-frontend/src/components/Browse.jsx (hunk 3 above)
Comment (antoniolinhart) [reply capstone-frontend/src/components/Browse.jsx]: Great idea! I didn't think of using a ternary here but it makes it much cleaner. I'll also replace it in the other 2 files

File: capstone-frontend/src/components/SearchBookModal.jsx (hunk 4 above)
Comment (antoniolinhart) [reply capstone-frontend/src/components/SearchBookModal.jsx]: Done!

File: capstone-frontend/src/components/CreateList.jsx (hunk 5 above)
Comment (antoniolinhart) [reply capstone-frontend/src/components/CreateList.jsx]: Done!

File: capstone-frontend/src/components/ClubSearchList.jsx (hunk 6 above)
Comment (antoniolinhart) [reply capstone-frontend/src/components/ClubSearchList.jsx:69]: With the original check, I kept running into the issue where this ClubSearchList page would reload without the new list of books being properly populated. It would render the page twice, once when the searchQuery changed and once when the fetches to the list of searched books finished.

For some reason, whenever I searched for a club then entered a query that shouldn't have resulted in any clubs, the list of clubs on this page stayed the same from the previous query.

This new check ensures that the page doesn't render multiple times and generates the correct clubs.

Comment (steven-solar) [reply capstone-frontend/src/components/ClubSearchList.jsx:69]: Got it, nice catch!

File: capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 9]
```diff
@@ -130,11 +130,13 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-      volumes.addAll(convertResponseToVolumeData(fullOutput));
-      if (volumes.size() == 0) {
-        response.sendError(HttpServletResponse.SC_NOT_FOUND);
+      // Check for empty response
+      Collection<VolumeData> volumesRetrieved = convertResponseToVolumeData(fullOutput);
+      if (volumesRetrieved == null || volumesRetrieved.size() == 0) {
+        response.sendError(HttpServletResponse.SC_NO_CONTENT);
         return;
       }
+      volumes.addAll(convertResponseToVolumeData(fullOutput));
```
Comment (VivekShankar) [inline capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java]: nit: Can you just add `volumesRetrieved` here?

**Review by VivekShankar — APPROVED** [review]
LGTM - nice error checking here!

File: capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java (hunk 9 above)
Comment (antoniolinhart) [reply capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java]: Yup! Good idea, I created the variable and replaced it in the first if check but not this one
